neuralop/models/codano.py

```diff
@@ -0,0 +1,371 @@
+                )
+            )
+        self.n_variables += num_variables
+
+    def _get_positional_encoding(self, x):
+        encoding_list = []
+        for i in range(self.n_variables):
+            encoding_list.append(torch.fft.irfftn(self.positional_encoding[i],
+                                                s=x.shape[-self.n_dim:]))
+        
+        return torch.stack(encoding_list, dim=1)
+    
+    def _get_cls_token(self, x):
+        cls_token = torch.fft.irfftn(self.cls_token, s=x.shape[-self.n_dim:])
+        repeat_shape = [1 for _ in x.shape]
+        repeat_shape[0] = x.shape[0]
+        cls_token = cls_token.repeat(*repeat_shape)
+        return cls_token
+    
+    def _extend_variables(self, x, static_channel):
+        x = x.unsqueeze(2)
+        if static_channel is not None:
+            repeat_shape = [1 for _ in x.shape]
+            repeat_shape[1] = x.shape[1]
+            static_channel = static_channel.unsqueeze(1).repeat(*repeat_shape)
+            x = torch.cat([x, static_channel], dim=2)
+        if self.use_positional_encoding:
+            positional_encoding = self._get_positional_encoding(x)
+            repeat_shape = [1 for _ in x.shape]
+            repeat_shape[0] = x.shape[0]
+            x = torch.cat([x, positional_encoding.repeat(*repeat_shape)], dim=2)
+        return x
+
+    def forward(self, x: torch.Tensor, static_channel=None):
+        '''
+        Parameters
+        ----------
+        x : torch.Tensor
+            input tensor of shape (batch_size, channel, H, W, ...)
+        static_channel : torch.Tensor
+            static channel tensor of shape (batch_size, static_channel_dim, H, W, ...)
+        '''
+        batch, n_variables, *spatial_shape = x.shape
+
+        if self.static_channel_dim > 0 and static_channel is None and static_channel.shape[1] != self.static_channel_dim:
+            raise ValueError("static channel dimension is not correct")
+        if self.use_positional_encoding and self.n_variables is None:
+            raise ValueError("number of variables is not given")
+        if self.use_positional_encoding:
+            assert x.shape[1]//self.n_variables == self.input_variable_codimension, "input variable codimension is not correct"
+        
+        x = self._extend_variables(x, static_channel) # (batch_size, n_variables, extended_codim, H, W, ...)
+
+        if self.lifting:
+            x = x.reshape(batch*n_variables, self.extended_variable_codimemsion, *spatial_shape)
+            x = self.lifting(x)
+            x = x.reshape(batch, n_variables*self.hidden_variable_codimension, *spatial_shape)
+
+        if self.enable_cls_token:
+            cls_token = self._get_cls_token(x)
+            x = torch.cat(
+                [
+                    cls_token,
+                    x,
+                ],
+                dim=1,
+            )
+
+        if self.domain_padding is not None:
+            x = self.domain_padding.pad(x)
+
+        output_shape = [
```

Comment:
Comments here explaining the different operations of the forward pass might be helpful too.

Reply:
Added